core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -219,7 +219,10 @@ final class Stream[+F[_],+O] private (private val coreRef: Stream.CoreRef[F,O])
   def pure(implicit S: Sub1[F,Pure]): Stream[Pure,O] = covary[Pure]
 
   /** Repeat this stream an infinite number of times. `s.repeat == s ++ s ++ s ++ ...` */
-  def repeat: Stream[F,O] = self ++ repeat
+  def repeat: Stream[F,O] = for {
+    a <- self.uncons1
+    b <- a.map { case (h, t) => t.cons1(h) ++ repeat }.getOrElse(Stream.empty)
```

Comment:
Also consider: eval_(f).repeat

It would be very strange if that only ran the effect once.

Sorry I don't think we can merge this, or any similar special casing.